src/wp-includes/media.php

```diff
@@ -5783,6 +5838,20 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
 		if ( wp_lazy_loading_enabled( $tag_name, $context ) ) {
 			$loading_attrs['loading'] = 'lazy';
 		}
+
+		// Do not override `decoding` if it is already set.
+		if ( 'img' === $tag_name ) {
+			$loading_attrs['decoding'] = 'async';
+			if ( array_key_exists( 'decoding', $attr ) ) {
+				$loading_attrs['decoding'] = $attr['decoding'];
+			}
+		}
+	}
+
+	// If the element is not in the viewport, and has not sizes, it should only have the decoding attribute.
+	if ( is_array( $maybe_size_less_attrs ) && isset( $loading_attrs['decoding'] ) ) {
+		$maybe_size_less_attrs['decoding'] = $loading_attrs['decoding'];
+		return $maybe_size_less_attrs;
```

Comment:
See above, this is unnecessarily complicated. We can instead apply the attribute logic much further up in the function, before all the `loading` and `fetchpriority` specific code.

On another note, please use `isset( $attr['decoding'] )` instead of `array_key_exists( 'decoding', $attr )` as it is a bit faster and can be used here without any drawbacks.